x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/async/AsyncTaskServiceTests.java

```diff
@@ -89,74 +92,108 @@ public void testEnsuredAuthenticatedUserIsSame() throws IOException {
+                ),
+            is(true)
         );
-        threadContext = new ThreadContext(Settings.EMPTY);
-        original.writeToContext(threadContext);
-        assertFalse(original.canAccessResourcesOf(differentRealmType));
-        assertFalse(indexService.ensureAuthenticatedUserIsSame(threadContext.getHeaders(), differentRealmType));
-
-        // wrong user
-        Authentication differentUser = new Authentication(
-            new User("test2", "role"),
-            new Authentication.RealmRef("realm", "realm", "node"),
-            null
-        );
-        assertFalse(original.canAccessResourcesOf(differentUser));
+
+        try (ThreadContext.StoredContext ignore = indexService.getSecurityContext().getThreadContext().stashContext()) {
+            // current user being run as
+            current = new Authentication(
+                user,
+                new Authentication.RealmRef(randomAlphaOfLengthBetween(1, 16), "file", "node"),
+                new Authentication.RealmRef("realm", "file", "node")
+            );
+            current.writeToContext(indexService.getSecurityContext().getThreadContext());
+            assertThat(
+                indexService.getSecurityContext().canIAccessResourcesCreatedWithHeaders(getAuthenticationAsHeaders(original)),
+                is(true)
+            );
+
+            // both users are run as
+            assertThat(
+                indexService.getSecurityContext()
+                    .canIAccessResourcesCreatedWithHeaders(
+                        getAuthenticationAsHeaders(
+                            new Authentication(
+                                user,
+                                new Authentication.RealmRef(randomAlphaOfLengthBetween(1, 16), "file", "node"),
+                                new Authentication.RealmRef("realm", "file", "node")
+                            )
+                        )
+                    ),
+                is(true)
+            );
+        }
+
+        try (ThreadContext.StoredContext ignore = indexService.getSecurityContext().getThreadContext().stashContext()) {
+            // different authenticated by type
+            Authentication differentRealmType = new Authentication(
+                new User("test", "role"),
+                new Authentication.RealmRef("realm", randomAlphaOfLength(10), "node"),
+                null
+            );
+            differentRealmType.writeToContext(indexService.getSecurityContext().getThreadContext());
+            assertFalse(indexService.getSecurityContext().canIAccessResourcesCreatedWithHeaders(getAuthenticationAsHeaders(original)));
+        }
+
+        // different user
+        try (ThreadContext.StoredContext ignore = indexService.getSecurityContext().getThreadContext().stashContext()) {
+            Authentication differentUser = new Authentication(
+                new User("test2", "role"),
+                new Authentication.RealmRef("realm", "realm", "node"),
```

Comment:
Since this assertion is to test different username, technically the RealmRef should be kept identical, i.e. 
```suggestion
                new Authentication.RealmRef("realm", "file", "node"),
```

Reply:
👍 thanks